modules/oe_subscriptions_anonymous/tests/src/FunctionalJavascript/SubscribeModalFormTest.php

```diff
@@ -0,0 +1,118 @@
+
+  /**
+   * Tests that subscribe link open a modal.
+   */
+  public function testModalForm(): void {
+    // Create an article content type.
+    $this->drupalCreateContentType([
+      'type' => 'article',
+      'name' => 'Article',
+    ]);
+    // Create a flag.
+    $flag = $this->createFlagFromArray([
+      'id' => 'subscribe_article',
+      'short_text' => 'Subscribe article',
+      'entity_type' => 'node',
+      'bundles' => ['article'],
+    ]);
+    // Create the node.
+    $node = $this->drupalCreateNode([
+      'type' => 'article',
+      'status' => 1,
+    ]);
+    $node->save();
+
+    $page = $this->getSession()->getPage();
+    $assert_session = $this->assertSession();
+    // Using ui-dialog parent of modal-form,
+    // Submit buttons and close in siblings.
+    $modal_selector = '.ui-dialog';
+
+    // Got to node page with subscription.
+    $this->drupalGet($node->toUrl());
+    // Click subscribe link.
+    $link_text = $flag->getShortText('flag');
+    $this->clickLink($link_text);
+    $assert_session->assertWaitOnAjaxRequest();
+
+    // The modal wrapper.
+    $modal = $page->find('css', $modal_selector);
+    // Find all elements.
+    $mail_label = 'Your e-mail';
+    $terms_label = 'I have read and agree with the data protection terms.';
+    $mail_field = $assert_session->fieldExists($mail_label, $modal);
+    $terms_field = $assert_session->fieldExists($terms_label, $modal);
+    // Get button pane with submit buttons.
+    $button_pane = $assert_session->elementExists('css', '.ui-dialog-buttonpane', $modal);
+
+    // Try to submit, empty form.
+    $this->disableNativeBrowserRequiredFieldValidation();
+    $assert_session->buttonExists('Subscribe me', $button_pane)->press();
+    $this->assertSession()->pageTextContains("$mail_label field is required.");
+    $this->assertSession()->pageTextContains("$terms_label field is required.");
+
+    // Test Close modal button.
+    $this->drupalGet($node->toUrl());
+    $this->clickLink($link_text);
+    $assert_session->assertWaitOnAjaxRequest();
+    $modal->find('css', 'button.ui-dialog-titlebar-close')->press();
+    $assert_session->elementNotExists('css', $modal_selector);
+    $assert_session->statusMessageNotExists();
+
+    // Test cancel button, filled data.
+    $this->clickLink($link_text);
+    $assert_session->assertWaitOnAjaxRequest();
+    $mail_field->setValue('test@test.com');
+    $terms_field->check();
+    $assert_session->buttonExists('No thanks', $button_pane)->press();
+    $assert_session->elementNotExists('css', $modal_selector);
+    $assert_session->statusMessageNotExists();
+
+    // Test submit.
+    $this->clickLink($link_text);
+    $assert_session->assertWaitOnAjaxRequest();
+    $mail_field->setValue('test@test.com');
+    $terms_field->check();
+    $assert_session->buttonExists('Subscribe me', $button_pane)->press();
+    $assert_session->elementNotExists('css', $modal_selector);
+    $assert_session->statusMessageExists();
```

Comment:
Now we need to test that the confirmation message is shown.

Reply:
Done